### PR TITLE
Kazuya rework

### PR DESCRIPTION
**Pull request by black-calculus**
(Currently compiling hitbox data, will update once finished)

### STATS:

- [+] Initial dash speed increased 1.775 → 1.8
- [+] Run speed increased 1.7 → 1.75
- [-] Stick-based additional run accel decreased 0.08323 → 0.08
- [+] Max horizontal jump speed increased 1.27 → 1.45

### Jab (10-Hit Combo):

- [R] Jab 1 and 2 hitboxes reworked
- Jab 1:
  - Startup decreased F6 → F4
  - Hitbox duration increased 2F → 3F
  - Angle 361 → 80
  - BKB 30 → 0
  - FKB 0 →25
  - KBG 10 → 100
- Jab 2:
  - Startup decreased F7 → F5
  - Hitbox duration increased 2F → 3F
  - Angle 361 → 75
  - BKB 30 → 0
  - FKB 0 →25
  - KBG 10 → 100

### Forward Tilt (Oni front Kick):

- [R] Now divided into sweetspot and sourspot
  - Sweetspot damage 14.5% → 15.0%; attached to foot and front knee
  - Sourspot damage 14.5% → 13.0%; attached to thigh and back knee
  - All other hitbox properties unchanged

### Up Tilt (Twin Pistons):

- [+] Hit 1 startup decreased F8 → F7
- [+] Endlag decreased by 3F; new FAF (combined with startup decrease) F30 → F26
- [+] Hit 2 BKB decreased 72 → 70
- [+] Hit 2 FAF decreased F34 → F33
- [*] Hit 2 shoulder hitbox damage normalized with all other hitboxes: 10.0% → 7.0%

### Down Tilt (Nejiri Uraken):

- [+] Startup decreased F12 → F10

### Back Tilt (Flash Tornado):

- [+] Startup decreased F11 → F10

### Up-Forward Tilt (Roundhouse to Triple Spin Kicks):

- [+] Startup decreased F16 → F11

### Down-Forward Tilt (Tsunami Kick)):

- [+] Hit 1 startup decreased F8 → F6
- [=] Hit 1 capsule hitboxes repositioned to better match animation
- [+] Hit 2 startup decreased F12 → F8
- [R] Hit 2 hitboxes reworked
  - Angle 50 → 115
  - BKB 65 → 70
  - KBG 70 → 65

### Up-Back Tilt (Jump Side Kick):

- [R] Hitboxes reworked; now separated into early and late hit
  - Early hit
    - F12-15
    - Damage 11.5% → 12.0%
  - Late hit
    - F16-17
    - Damage 11.5% → 10.0%

### Down-Back Tilt (Stature Smash):

- [+] Startup decreased F13 → F11
- [+] Endlag decreased by 1F

### Crouch Jab:

- [!] Can now properly jab lock opponents

### Crouching Back Tilt (Crouching Spin Kick):

- [R] BKB decreased 90 → 80
- [R] KBG increased 10 → 20
- [R] Angle increased 55 -> 78

### While Rising Attack (Demon God Fist):

- [R] BKB decreased 60 → 40
- [R] Angle increased 361 → 65
- [*] Hitbox ID 2 KBG normalized with all other hitboxes

### Forward Smash (Glorious Demon God Fist):

- [+] Hitbox duration increased 3F → 4F
- [-] Sourspot damage decreased 23% → 20%
- [=] Sourspot KBG increased 63 → 65
- [-] Sweetspot damage decreased 26% → 25%
- [=] Sweetspot KBG increased 73 → 75
- [+] Sourspot hitboxes added to shoulder and arm for better coverage

### Up Smash (Devil Twister):

- [$] Hitlag multiplier increased 0.3x → 0.75x
- [+] Late hit hitboxes now match early hit
- [+] Hitbox duration increased F14-18 → F14-19

### Down Smash (Lion Slayer):

- [$] Hitlag multiplier increased 0.3x → 0.75x

### Standing Grab:
- [+] Air-only and ground-only hitboxes merged, size now matches old ground-only hitbox

### Back Throw (Steel Pedal Drop):

- [!] Now knocks opponents down
- [+] FAF decreased F83 → F62

### Down Throw (Stonehead):

- [+] KBG decreased 140 → 135
- [+] Angle increased 77 → 80

### Neutral Air (Jumping Knuckle):

- [R] Early hit fist sourspot removed
- [+] Early hit fist sweetspot hitbox radius increased to match old sourspot size (3.0u → 4.5u)
- [+] Sourspot hitbox added to shoulder
- [$] Hitlag multiplier increased 0.65x → 0.75x

### Forward Air (Searing Edge):

- [+] Early hit sourspot BKB decreased 42 → 40
- [$] Early hit hitlag multiplier increased 0.65x → 0.75x
- [+] Late hit KBG decreased 71 → 70

### Up Air (Rising Toe Kick):

- [+] Early hit sourspot angle increased 70 → 75
- [+] Early hit sourspot KBG decreased 75 → 70
- [+] Late hit sourspot angle increased 70 → 75
- [+] Late hit sourspot BKB increased 40 → 50
- [+] Late hit sourspot KBG decreased 75 → 70

### Devil Fist:

 - [+] Kazuya can now perform side special and up special in the same airtime again (though you only get one side special per airtime; does not reset on hit)

### Devil Wings:

- [$] Early hit sweetspot hitlag multiplier increased 0.3x → 0.8x
- [$] Early hit sourspot hitlag multiplier increased 0.3x → 0.5x
- [$] Mid hit hitlag multiplier increased 0.3x → 0.5x
- [$] Late hit hitlag multiplier increased 0.2x → 0.5x

### Heaven's Door:

- [+] Air-only and ground-only hitboxes merged, size now matches old ground-only hitbox
- [+] Aerial FAF decreased F60 → F55

### Wind God Fist:

- [+] FAF decreased F36 → F33

### Spinning Demon to Left Hook:

- [+] Startup decreased F10 → F8
- [+] SDI multiplier decreased 0.8x → 0.0x
- [+] Hit 1 additional hitstun frames increased 12F → 20F
- [+] Hit 2 angle decreased 35 → 25
- [+] Hit 2 BKB decreased 85 → 79
- [+] Hit 2 KBG decreased 24 → 20
- [-] Hit 2 additional hitstun modifier decreased +8F → -10F

### OTHER:
- [*] Fixed an issue where Kazuya's aerials weren't properly retaining jump momentum

### CODE REVIEW

File: romfs/source/fighter/common/param/fighter_param.prcxml [hunk 1]
```diff
@@ -2664,13 +2664,13 @@
...
-      <float hash="dash_speed">1.775</float>
-      <float hash="run_accel_mul">0.08323</float>
+      <float hash="dash_speed">1.8</float>
+      <float hash="run_accel_mul">0.08</float>
       <float hash="run_accel_add">0.02</float>
-      <float hash="run_speed_max">1.7</float>
+      <float hash="run_speed_max">1.75</float>
       <int hash="jump_squat_frame">5</int>
       <float hash="jump_speed_x">1</float>
-      <float hash="jump_speed_x_max">1.27</float>
```
Comment (SuddyN) [inline romfs/source/fighter/common/param/fighter_param.prcxml:2673]: listed as 1.28 --> 1.45, but is actually 1.27 --> 1.45

File: fighters/demon/src/acmd/jabs.rs [hunk 2]
```diff
@@ -10,24 +10,32 @@ unsafe fn game_attack11(fighter: &mut L2CAgentBase) {
...
+        ATTACK(fighter, 2, 0, Hash40::new("arml"), 3.0, 80, 100, 25, 0, 3.0, 4.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA_d, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
+        ATTACK(fighter, 3, 0, Hash40::new("top"), 3.0, 80, 100, 25, 0, 4.0, 0.0, 10.0, 4.0, Some(0.0), Some(10.0), Some(9.0), 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA_d, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/jabs.rs:23]: angle listed as 75 but is actually 80

File: fighters/demon/src/acmd/jabs.rs [hunk 3]
```diff
@@ -10,24 +10,32 @@ unsafe fn game_attack11(fighter: &mut L2CAgentBase) {
...
+        // Jab lock hitbox
+        ATTACK(fighter, 4, 0, Hash40::new("top"), 3.0, 361, 10, 0, 30, 3.0, 0.0, 4.0, 3.0, Some(0.0), Some(4.0), Some(9.0), 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_G, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
+        AttackModule::set_down_only(boma, 4, false);
+
+        AttackModule::set_add_reaction_frame_revised(boma, 0, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 1, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 2, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 3, 2.0, false);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/jabs.rs:31]: I don't like using add_reaction_frame, but Kazuya is weird so you can dismiss if you want

File: fighters/demon/src/acmd/jabs.rs [hunk 4]
```diff
@@ -44,27 +52,34 @@ unsafe fn game_attack12(fighter: &mut L2CAgentBase) {
...
+        ATTACK(fighter, 2, 0, Hash40::new("arml"), 3.0, 75, 100, 25, 0, 3.0, 4.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA_d, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
+        ATTACK(fighter, 3, 0, Hash40::new("top"), 3.0, 75, 100, 25, 0, 4.0, 0.0, 10.0, 4.0, Some(0.0), Some(10.0), Some(9.0), 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA_d, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/jabs.rs:64]: angle change not listed in changelog

File: fighters/demon/src/acmd/jabs.rs [hunk 5]
```diff
@@ -44,27 +52,34 @@ unsafe fn game_attack12(fighter: &mut L2CAgentBase) {
...
+        // Jab lock hitbox
+        ATTACK(fighter, 4, 0, Hash40::new("top"), 3.0, 361, 10, 0, 30, 3.0, 0.0, 4.0, 3.0, Some(0.0), Some(4.0), Some(9.0), 1.0, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 0, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_G, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new_raw(0x193bdcb0cc), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
+        AttackModule::set_down_only(boma, 4, false);
+
+        AttackModule::set_add_reaction_frame_revised(boma, 0, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 1, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 2, 2.0, false);
+        AttackModule::set_add_reaction_frame_revised(boma, 3, 2.0, false);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/jabs.rs:72]: I don't like using add_reaction_frame (especially for jabs), but Kazuya is weird so you can dismiss if you want

File: fighters/demon/src/acmd/jabs.rs [hunk 6]
```diff
@@ -299,8 +314,10 @@ unsafe fn game_flashpunch(fighter: &mut L2CAgentBase) {
...
             ATTACK(fighter, 1, 1, Hash40::new("top"), 0.0, 368, 100, 40, 0, 7.0, 0.0, 15.0, 0.0, Some(0.0), Some(7.0), Some(0.0), 0.0, 0.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 0, 0.0, 0, false, false, true, true, false, *COLLISION_SITUATION_MASK_A, *COLLISION_CATEGORY_MASK_FIGHTER, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_NONE, *ATTACK_REGION_NONE);
-            AttackModule::set_vec_target_pos(boma, 0, smash::phx::Hash40::new("top"), &Vector2f::new(5.0, 15.0), 5, false);
-            AttackModule::set_vec_target_pos(boma, 1, smash::phx::Hash40::new("top"), &Vector2f::new(5.0, 15.0), 5, false);
+            let x_pos = 13.0*PostureModule::lr(boma);
+            let y_pos = 6.0;
+            AttackModule::set_vec_target_pos(boma, 0, smash::phx::Hash40::new("top"), &Vector2f::new(x_pos, y_pos), 4, false);
+            AttackModule::set_vec_target_pos(boma, 1, smash::phx::Hash40::new("top"), &Vector2f::new(x_pos, y_pos), 4, false);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/jabs.rs]: doesn't AttackModule already take care of flipping these vec_targets for us? Won't this flip angle when facing left vs right? What even is game_flashpunch anyway, can you add a comment above this script to denote it?

File: fighters/demon/src/acmd/tilts.rs [hunk 7]
```diff
@@ -48,6 +49,7 @@ unsafe fn game_attackhi3(fighter: &mut L2CAgentBase) {
         }
     }
     frame(lua_state, 4.0);
+    FT_MOTION_RATE_RANGE(fighter, 4.0, 8.0, 3.0);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/tilts.rs:52]: there's no line to reset the motion rate on this, until frame 21 (seems like it should be frame 8) -- this actually shaves off 4.25 frames from the move (25% of the frames between 4 and 21)

File: fighters/demon/src/acmd/tilts.rs [hunk 8]
```diff
@@ -78,6 +80,7 @@ unsafe fn game_attackhi3(fighter: &mut L2CAgentBase) {
         WorkModule::on_flag(boma, *FIGHTER_DEMON_STATUS_ATTACK_HI_3_FLAG_CHECK_STEP);
     }
     frame(lua_state, 21.0);
+    FT_MOTION_RATE_RANGE(fighter, 21.0, 30.0, 6.0);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/tilts.rs:84]: more frames cut off the end of the move with no reset in motion rate. In your changelog, UTilt1 has no changes listed.

File: fighters/demon/src/acmd/tilts.rs [hunk 9]
```diff
@@ -110,10 +113,10 @@ unsafe fn game_attackhi32(fighter: &mut L2CAgentBase) {
...
+            ATTACK(fighter, 2, 0, Hash40::new("armr"), 7.0, 83, 64, 0, 70, 5.0, 0.0, 0.0, 0.0, None, None, None, 0.65, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 1, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL_M, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
+            ATTACK(fighter, 3, 0, Hash40::new("shoulderr"), 7.0, 83, 64, 0, 70, 3.0, 0.0, 0.0, 0.0, None, None, None, 0.65, 1.0, *ATTACK_SETOFF_KIND_ON, *ATTACK_LR_CHECK_POS, false, 1, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL_M, *COLLISION_SOUND_ATTR_DEMON_PUNCH01, *ATTACK_REGION_PUNCH);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/tilts.rs:120]: removal of this sweetspot (dmg 10 --> 7) is not listed in the changelog

File: fighters/demon/src/acmd/smashes.rs [hunk 10]
```diff
@@ -31,22 +32,23 @@ unsafe fn demon_attack_s4_game(fighter: &mut L2CAgentBase) {
     FT_MOTION_RATE(fighter, 1.0);
     frame(lua_state, 25.0);
     if is_excute(fighter) {
+        FT_DESIRED_RATE(fighter, 28.0-25.0, 4.0);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/smashes.rs]: line 35 `FT_DESIRED_RATE` should be removed and replace line 32 `FT_MOTION_RATE` (all motion rate changes should go outside is_excute blocks)

File: fighters/demon/src/acmd/smashes.rs [hunk 11]
```diff
@@ -65,14 +67,14 @@ unsafe fn demon_attack_s4_game(fighter: &mut L2CAgentBase) {
...
-unsafe fn demon_attack_hi4_game(fighter: &mut L2CAgentBase) {
+unsafe fn game_attackhi4(fighter: &mut L2CAgentBase) {
     let lua_state = fighter.lua_state_agent;
     let boma = fighter.boma();
     smash::app::FighterSpecializer_Demon::set_devil(boma, true, 10.0);
     frame(lua_state, 1.0);
-    FT_MOTION_RATE(fighter, 0.7);
     if is_excute(fighter) {
         JostleModule::set_team(boma, 1);
+        FT_DESIRED_RATE(fighter, 14.0-1.0, 9.0);
```
Comment (SuddyN) [inline fighters/demon/src/acmd/smashes.rs]: move this up two lines (all motion rate changes should go outside is_excute blocks)

File: fighters/demon/src/acmd/throws.rs [hunk 12]
```diff
@@ -26,6 +47,62 @@ unsafe fn demon_throw_hi_expression(fighter: &mut L2CAgentBase) {
     }
 }
 
+#[acmd_script( agent = "demon", script = "game_throwb", category = ACMD_GAME, low_priority )]
+unsafe fn game_throwb(fighter: &mut L2CAgentBase) {
+    let lua_state = fighter.lua_state_agent;
+    let boma = fighter.boma();
+    if !smash2::app::FighterCutInManager::is_vr_mode() {
```
Comment (SuddyN) [inline fighters/demon/src/acmd/throws.rs:54]: VR support 😭 (dismiss)

**Review by SuddyN — DISMISSED** [review]

File: fighters/demon/src/acmd/jabs.rs (hunk 6 above)
Comment (black-calculus) [reply fighters/demon/src/acmd/jabs.rs]: Was having some very strange issues using `set_vec_target_pos()` for this move, for some reason it wasn't changing the relative position with his facing direction taken into account so I had to do that manually. Not sure why, can't determine anything about this move that would differentiate it from the other uses of the function but 🤷‍♂️

Comment (WuBoytH) [reply fighters/demon/src/acmd/jabs.rs]: I believe it's because you have `ATTACK_LR_CHECK_F`, try `POS` instead

**Review by blu-dev — APPROVED** [review]
gottem ggs

**Pull request merged**